mlx/transforms.cpp

```diff
@@ -216,13 +218,43 @@ void eval(const std::vector<array>& outputs) {
   auto synchronizer =
       array({}, bool_, std::make_unique<Synchronizer>(stream), outputs);
 
-  recurse = [&](const array& a) {
+  compute_depths = [&](const array& a) {
+    auto id = a.id();
+    if (auto it = depth_map.find(id); it != depth_map.end()) {
+      return it->second;
+    }
+    int depth = 0;
+    for (auto& in : a.inputs()) {
+      depth = std::max(depth, compute_depths(in));
+    }
+    depth++;
+    depth_map[id] = depth;
+    for (auto& s : a.siblings()) {
+      depth_map[s.id()] = depth;
+    }
+    return depth;
```

Comment:
Did you consider trying to count the aggregate array size of the branch rather than the depth?

Reply:
This may indeed be a more robust heuristic. I think we should consider it as a TODO though cause the current one works and we can keep at it for a while trying new things without much benefit...